image-rauc.c

```diff
@@ -105,10 +107,21 @@ static int rauc_generate(struct image *image)
 				goto out;
 		}
 
-		image_info(image, "adding file '%s' as '%s' ...\n",
-				child->file, target);
-		ret = systemp(image, "cp --remove-destination '%s' '%s/%s'",
-				file, tmpdir, target);
+		xasprintf(&tmptarget, "%s/%s", tmpdir, target);
+
+		image_info(image, "adding file '%s' as '%s' (offset=%lld)...\n",
+				child->file, target, (long long)part->imageoffset);
+
+		unlink(tmptarget);
+
+		/*
+		 * Starting with coreutils 9.1 you can use a 'B' suffix for
+		 * skip=N instead of iflag=skip_bytes to have N count bytes, not
+		 * (input) blocks.
+		 */
+		ret = systemp(image, "dd if='%s' of='%s' conv=sparse iflag=skip_bytes skip=%lld",
+				file, tmptarget, (long long)part->imageoffset);
```

Comment:
This does not work in all cases. In general, we assume, that explicit holes are 'don't care' areas, because they are created when a mkfs skips the area. And mkfs never assumes that the underlying device contains zeros.

And we make use of that in the android sparse image: holes are turned into 'don't care' blocks and zeros are turned into 'fill' blocks.

But `dd` breaks this: it turns all blocks that are zero into holes and does not care if the input files has holes  or not. `cp` on the other hand propagates holes explicitly by default, so that's save.

So please keep the `cp` for `offset == 0` and drop the `conv=sparse` for the other case. It's not the best solution but the main use-case for the offset is bootloader images and those typically don't contain holes anyways.

A better solution would be to do this manually, but `insert_image()` needs to be extended to support the offset first. We can do that later.